File: softphone/src/main/java/sip/softphone/JainFactories.java
package sip.softphone;

import sip.softphone.exceptions.JainWrappedException;

import javax.inject.Inject;
import javax.inject.Singleton;
import javax.sip.*;
import javax.sip.address.AddressFactory;
import javax.sip.header.HeaderFactory;
import javax.sip.message.MessageFactory;
import java.util.Properties;

@Singleton
public class JainFactories {
    private final Object factoryLock;
    private final SipStackDetails stackDetails;
    private SipFactory createdFactory;

    @Inject
    public JainFactories(final SipStackDetails stackDetails) {
        this.stackDetails = stackDetails;
        factoryLock = new Object();
    }

    public SipFactory getSipFactory() {
        synchronized (factoryLock) {
            if (createdFactory == null) {
                createdFactory = createSipFactory();
            }
        }
        return createdFactory;
    }

    private SipFactory createSipFactory() {
        SipFactory factory = SipFactory.getInstance();
        factory.setPathName("gov.nist");
        return factory;
    }

    public HeaderFactory headerFactory() throws JainWrappedException {
        try {
            return getSipFactory().createHeaderFactory();
        } catch (PeerUnavailableException e) {
            throw new JainWrappedException("headerFactory failed with PeerUnavailable exception", e);
        }
    }

    public MessageFactory messageFactory() throws JainWrappedException {
        try {
            return getSipFactory().createMessageFactory();
        } catch (PeerUnavailableException e) {
            throw new JainWrappedException("messageFactory failed with PeerUnavailable exception", e);
        }
    }

    public AddressFactory addressFactory() throws JainWrappedException {
        try {
            return getSipFactory().createAddressFactory();
        } catch (PeerUnavailableException e) {
            throw new JainWrappedException("addressFactory failed with PeerUnavailable exception", e);
        }
    }

    public SipStack createSipStack() throws JainWrappedException {
        Properties properties = new Properties();
        properties.setProperty("javax.sip.STACK_NAME", stackDetails.stackName());
        properties.setProperty("javax.sip.IP_ADDRESS", stackDetails.ipAddress());

        try {
            return getSipFactory().createSipStack(properties);
        } catch (PeerUnavailableException e) {
            throw new JainWrappedException("createSipStack failed with PeerUnavailable exception", e);
        }
    }

    public SipProvider createSipProvider() throws JainWrappedException {
        return createSipProviderWithStackAndDetails(createSipStack(), stackDetails);
    }

    public SipProvider createSipProviderWithStackDetails(final SipStackDetails stackDetails) throws JainWrappedException {
        return createSipProviderWithStackAndDetails(createSipStack(), stackDetails);
    }

    public SipProvider createSipProviderWithStackAndDetails(final SipStack sipStack, final SipStackDetails stackDetails) throws JainWrappedException {
        ListeningPoint listeningPoint = null;
        try {
            listeningPoint = sipStack.createListeningPoint(stackDetails.ipAddress(), stackDetails.port(), stackDetails.protocol());
        } catch (TransportNotSupportedException e) {
            throw new JainWrappedException("createListeningPoint had a TransportNotSupported exception", e);
        } catch (InvalidArgumentException e) {
            throw new JainWrappedException("createListeningPoint had an InvalidArgument exception", e);
        }

        try {
            return sipStack.createSipProvider(listeningPoint);
        } catch (ObjectInUseException e) {
            throw new JainWrappedException("createSipProvider failed with ObjectInUse Exception", e);
        }
    }
}

File: softphone/src/main/kotlin/sip/softphone/factories/JainFactories.kt
package sip.softphone.factories

import sip.softphone.SipStackDetails
import sip.softphone.exceptions.JainWrappedException
import java.util.*
import javax.inject.Inject
import javax.sip.*
import javax.sip.address.AddressFactory
import javax.sip.header.HeaderFactory
import javax.sip.message.MessageFactory

open class JainFactories @Inject constructor(val sipStackDetails: SipStackDetails) {
    val sipFactory: SipFactory by lazy {
        createSipFactory()
    }

    private fun createSipFactory(): SipFactory {
        val factory = SipFactory.getInstance()
        factory.pathName = "gov.nist"
        return factory
    }

    open fun headerFactory(): HeaderFactory {
        try {
            return sipFactory.createHeaderFactory()
        } catch (e: PeerUnavailableException) {
            throw JainWrappedException("headerFactory failed with PeerUnavailable exception", e)
        }
    }

    open fun messageFactory(): MessageFactory {
        try {
            return sipFactory.createMessageFactory()
        } catch (e: PeerUnavailableException) {
            throw JainWrappedException("messageFactory failed with PeerUnavailable exception", e)
        }
    }

    open fun addressFactory(): AddressFactory {
        try {
            return sipFactory.createAddressFactory()
        } catch (e: PeerUnavailableException) {
            throw JainWrappedException("addressFactory failed with PeerUnavailable exception", e)
        }
    }

    open fun createSipStack(): SipStack {
        val properties = Properties()
        properties.setProperty("javax.sip.STACK_NAME", sipStackDetails.stackName())
        properties.setProperty("javax.sip.IP_ADDRESS", sipStackDetails.ipAddress())

        try {
            return sipFactory.createSipStack(properties)
        } catch (e: PeerUnavailableException) {
            throw JainWrappedException("createSipStack failed with PeerUnavailable exception", e)
        }
    }

    open fun createSipProvider(): SipProvider {
        return createSipProviderWithStackAndDetails(createSipStack(), sipStackDetails)
    }

    open fun createSipProviderWithStackDetails(stackDetails: SipStackDetails): SipProvider {
        return createSipProviderWithStackAndDetails(createSipStack(), stackDetails)
    }

    open fun createSipProviderWithStackAndDetails(sipStack: SipStack, stackDetails: SipStackDetails): SipProvider {
        val listeningPoint: ListeningPoint = try {
            sipStack.createListeningPoint(stackDetails.ipAddress(), stackDetails.port(), stackDetails.protocol())
        } catch (e: TransportNotSupportedException) {
            throw JainWrappedException("createListeningPoint had a TransportNotSupported exception", e)
        } catch (e: InvalidArgumentException) {
            throw JainWrappedException("createListeningPoint had an InvalidArgument exception", e)
        }

        try {
            return sipStack.createSipProvider(listeningPoint)
        } catch (e: ObjectInUseException) {
            throw JainWrappedException("createSipProvider failed with ObjectInUse Exception", e)
        }
    }
}